flow/MkCertCli.cpp

```diff
@@ -0,0 +1,326 @@
+	printOptionUsage("--expire-client-cert (default: no)",
+	                 { "Deliberately expire client's leaf certificate for testing." });
+	printOptionUsage("--print-server-cert (default: no)",
+	                 { "Print generated server certificate chain including root in human readable form.",
+	                   "Printed certificates are in leaf-to-CA order.",
+	                   "If --print-client-cert is also used, server chain precedes client's." });
+	printOptionUsage("--print-client-cert (default: no)",
+	                 { "Print generated client certificate chain including root in human readable form.",
+	                   "Printed certificates are in leaf-to-CA order.",
+	                   "If --print-server-cert is also used, server chain precedes client's." });
+	printOptionUsage("--print-args (default: no)", { "Print chain generation arguments." });
+}
+
+struct ChainSpec {
+	unsigned length;
+	std::string certFile;
+	std::string keyFile;
+	std::string caFile;
+	mkcert::ESide side;
+	bool expireLeaf;
+	void transformPathToAbs() {
+		certFile = abspath(certFile);
+		keyFile = abspath(keyFile);
+		caFile = abspath(caFile);
+	}
+	void print() {
+		fmt::print(stdout, "{}-side:\n", side == mkcert::ESide::Server ? "Server" : "Client");
+		fmt::print(stdout, "  Chain length: {}\n", length);
+		fmt::print(stdout, "  Certificate file: {}\n", certFile);
+		fmt::print(stdout, "  Key file: {}\n", keyFile);
+		fmt::print(stdout, "  CA file: {}\n", caFile);
+		fmt::print(stdout, "  Expire cert: {}\n", expireLeaf);
+	}
+	mkcert::CertChainRef makeChain(Arena& arena);
+};
+
+mkcert::CertChainRef ChainSpec::makeChain(Arena& arena) {
+	auto checkStream = [](std::ofstream& fs, std::string_view filename) {
+		if (!fs) {
+			throw std::runtime_error(fmt::format("cannot open '{}' for writing", filename));
+		}
+	};
+	auto ofsCert = std::ofstream(certFile, std::ofstream::out | std::ofstream::trunc);
+	checkStream(ofsCert, certFile);
+	auto ofsKey = std::ofstream(keyFile, std::ofstream::out | std::ofstream::trunc);
+	checkStream(ofsKey, keyFile);
+	auto ofsCa = std::ofstream(caFile, std::ofstream::out | std::ofstream::trunc);
+	checkStream(ofsCa, caFile);
+	if (!length)
+		return {};
+	auto specs = mkcert::makeCertChainSpec(arena, length, side);
+	if (expireLeaf) {
+		specs[0].offsetNotBefore = -60l * 60 * 24 * 365;
+		specs[0].offsetNotAfter = -10l;
+	}
+	auto chain = mkcert::makeCertChain(arena, specs, {} /*generate root CA*/);
+	auto ca = chain.back().certPem;
+	ofsCa.write(reinterpret_cast<char const*>(ca.begin()), ca.size());
+	auto chainMinusRoot = chain;
+	if (chainMinusRoot.size() > 1)
+		chainMinusRoot.pop_back();
+	auto cert = mkcert::concatCertChain(arena, chainMinusRoot);
+	ofsCert.write(reinterpret_cast<char const*>(cert.begin()), cert.size());
+	auto key = chain[0].privateKeyPem;
+	ofsKey.write(reinterpret_cast<char const*>(key.begin()), key.size());
+	ofsCert.close();
```

Comment:
The file streams should be closed automatically by the destructors

Reply:
Yes, I am aware. I thought it would be better to be explicit about it.

Reply:
Then leave it so. I just thought because there is a return in the middle that relies on the destructors.